ecs-cli/modules/clients/aws/servicediscovery/client.go

```diff
@@ -0,0 +1,139 @@
+// Copyright 2015-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+// Package servicediscovery contains functions for working with the route53 APIs
+// that back ECS Service Discovery
+package servicediscovery
+
+import (
+	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/clients"
+	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/config"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/route53"
+	"github.com/aws/aws-sdk-go/service/servicediscovery"
+)
+
+// FindPrivateNamespace returns the ID(s) of the private namespace with the given name and vpc
+func FindPrivateNamespace(name string, vpc string, params *config.CommandConfig) ([]*string, error) {
+	region := aws.StringValue(params.Session.Config.Region)
+	var namespacesWithName []*string
+	err := listNamespaces(false, params, func(namespace *servicediscovery.NamespaceSummary) bool {
+		if aws.StringValue(namespace.Name) == name {
+			namespacesWithName = append(namespacesWithName, namespace.Id)
+		}
+		return true
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	var matchingNamespaces []*string
+	for _, namespaceID := range namespacesWithName {
+		hasVPC, err := checkNamespaceVPC(namespaceID, vpc, region, params)
+		if err != nil {
+			return nil, err
+		}
+		if hasVPC {
+			matchingNamespaces = append(matchingNamespaces, namespaceID)
+		}
+	}
+
+	return matchingNamespaces, nil
+}
+
+func checkNamespaceVPC(namespaceID *string, vpc string, region string, params *config.CommandConfig) (bool, error) {
+	namespaceInfo, err := getNamespace(namespaceID, params)
+	if err != nil {
+		return false, err
+	}
+	hostedZoneID := namespaceInfo.Namespace.Properties.DnsProperties.HostedZoneId
+	hostedZone, err := getHostedZone(hostedZoneID, params)
+	if err != nil {
+		return false, err
+	}
+	for _, hostedZoneVPC := range hostedZone.VPCs {
+		// The VPC must be in the region that we will be launching the ECS Service
+		if (aws.StringValue(hostedZoneVPC.VPCId) == vpc) && (aws.StringValue(hostedZoneVPC.VPCRegion) == region) {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// ServiceDiscovery API Actions that can be mocked in unit tests
+var sdClient *servicediscovery.ServiceDiscovery
```

Comment:
I appreciate what you're doing here to avoid messy client mocking, but I'm wondering if there's a better way to access clients since they seem to be instantiated within private function calls a lot.

Reply:
Each client is only created once per execution of the CLI, since the private functions get them if they have already been created.

Reply:
I'm a bit confused by the question- are you asking about why the SDK API calls happen in private functions which are set to vars? That is to make the API calls easy to mock. The logic is that its easy to mock a few API calls vs generating an entire mock object which mocks things that we do not even use. Please let me know if this is not what you were asking about.

Reply:
What I was trying to say is that this approach makes any interaction with the SD or Route53 much less extensible if we are wrapping each API call individually in a private function that also contains other business logic. I agree it's easy now since we're only using a limited set of API calls, and I am willing to concede I am prematurely anticipating an abstraction here. I believe we can still have a easily mockable client objects that takes methods that translate to API calls, but  I'm open to seeing how maintainable this approach is.